torchaudio/models/wav2vec2/model.py

```diff
@@ -137,18 +137,27 @@ class HuBERTPretrainModel(Module):
 
         logit_generator (torch.nn.Module):
             Logit generator that predicts the logits of the masked and unmasked inputs.
+
+        feature_grad_mult (float):
+            The factor to scale the convolutional feature extraction layer gradients by.
+            The scale factor will not affect the forward pass.
     """
 
     def __init__(
         self,
         wav2vec2: Wav2Vec2Model,
         mask_generator: Module,
         logit_generator: Module,
+        feature_grad_mult: Optional[float],
     ):
         super().__init__()
         self.wav2vec2 = wav2vec2
         self.mask_generator = mask_generator
         self.logit_generator = logit_generator
+        assert (
+            0.0 <= feature_grad_mult <= 1.0
+        ), f"The value of `feature_grad_mult` must be between [0, 1]. Found {feature_grad_mult}"
```

Comment:
I think it should be exclusive in both end.
It does not make much sense if it is 0 or 1.

Reply:
It makes sense when it is 0. In fine-tuning case, the parameters of feature extractor are fixed, which can be done by setting it to 0

Reply:
I think in such cases, freezing the entire model using PyTorch's API is more appropriate. That way, the backward path is stopped. Relying on `feature_grad_mult =0` means that backward path is still computed.

Reply:
There are two stages in the fine-tuning, in the first stage, only `aux` layer has gradient, the rest of parameters are fixed. In the second stage, the feature extractor's parameters are fixed. The train loop looks like:
```python
if self.global_step <= self.fix_encoder_updates:
    with torch.no_grad():
        x, out_len = self.model.wav2vec2.feature_extractor(waveforms, audio_lengths)
        padding_mask = components._get_padding_mask(x, out_len)
        x, attention_mask = self.model.wav2vec2.encoder._preprocess(x, out_len)
        x, mask = self.model.mask_generator(x, padding_mask)
        x = self.model.wav2vec2.encoder.transformer(x, attention_mask=attention_mask)
else:
    with torch.no_grad():
        x, out_len = self.model.wav2vec2.feature_extractor(waveforms, audio_lengths)
    padding_mask = components._get_padding_mask(x, out_len)
    x, attention_mask = self.model.wav2vec2.encoder._preprocess(x, out_len)
    x, mask = self.model.mask_generator(x, padding_mask)
    x = self.model.wav2vec2.encoder.transformer(x, attention_mask=attention_mask)
logits = self.aux(x)
```
Is there a better way to simplify the logic?

Reply:
> Is there a better way to simplify the logic?

Since the data flow is same, I think one can set `requires_grad` outside of the loop when it hits the right condition.

The pseudo code would look like;
1. Initialize the `self.model.wav2vec2`
2. Disable gradient entirely
3. Run the training loop for `self.fix_encoder_updates`
4. Once the `self.fix_encoder_updates` step is done, enable gradient for `self.model.wav2vec2.feature_extractor`.

Reply:
I see. I need to search where is the best place to enable gradient. `train_step` is called every iteration, so it's not suitable to call gradient enabling here.

to calrify, during the whole fine-tuning, `self.model.wav2vec2.feature_extractor` is freezed, otherwise the training will be unstable. In the first `fix_encoder_updates` iterations, the wav2vec2 model is freezed, only `aux` layer has gradient. after that, the gradient of `wav2vec2.encoder` is enabled.